text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+
+* There are no peers advertising the maximum known difficulty on the network also advertising the `PIBD_HIST_1` capability
+* The node has not seen such a peer for 60 seconds
+
+#### Requesting output bitmap segments
+
+Before any other segments can be validated or applied to local MMR storage, the syncing node must first request and validate the complete output bitmap set up to the MMR size represented in the horizon header.
+
+The MMR size, number of leaves and required number of bitmap segments can be pre-determined based on the size of the output MMR in the horizon header as demonstrated by the following reference code snippet:
+
+```
+// Calculate number of leaves in the bitmap mmr
+bitmap_mmr_leaf_count = (pmmr::n_leaves(self.archive_header.output_mmr_size) + 1023) / 1024;
+
+// Total size of Bitmap PMMR
+bitmap_mmr_size = 1 + pmmr::peaks(pmmr::insertion_to_pmmr_index(self.bitmap_mmr_leaf_count)).last()
+```
+
+The node should request bitmap segments from other peers according to the [Peer Selection Strategy](#peer-selection) above. (The process of re-creating an MMR from segments is described below TBD). Once the bitmap MMR is reconstructed ([process described below](#applying-segments-to-mmrs),) the node should keep a representation of the underlying bitmap cached in order to facilitate comparison against later incoming segments.
+
+The syncing node is not required to store the contents of the output set bitmap locally. If the PIBD process is interrupted, the node may re-request all bitmap segments from peers as part of the normal process of starting/resuming PIBD.
+
+#### Determining starting position for each MMR
+
+Once the bitmap set has been recreated and the node is ready to start requesting segments from peers, the node should examine the contents of its local MMR storage to determine the on-disk sizes of each PMMR. Based on this, it should determine which segments it should request first for each MMR. This ensures that the PIBD process is the same regardless of whether the node is starting from the genesis block, resuming, or catching-up after an extended period of being offline.
+
+#### Requesting Segments
+
+Once the required list of segments has been derived, the node can then begin to request segments in parallel from peers according to the [peer selection strategy](#peer-selection) above. Exactly how the node does this is left as an implementation detail, but the following guidelines should be kept in mind:
+
+* MMRs are an append-only structure, and can only be reconstucted sequentially left-to-right. Therefore there is little point requesting too many segments for a given MMR in advance, particularly if memory is limited.
+* A reasonable effort should be made to spread requests across the 3 UTXO MMRs to allow their reconstruction to complete around the same time.
```

Comment:
What's the benefit of having the three MMRs complete at the same time, over doing them purely in sequence for example?

Reply:
I'd originally had the notion of being able to validate the entire tx set up to a given 'block', i.e. all PMMRs are complete up to a point representing the application of a particular block. In this case, clients would be able to validate all objects and update the chain head accordingly, then when the node is killed and restarted it can resume from that point. In practice I don't think this is a huge benefit (other than being able to update the chain head as we go), since nodes can start resuming MMRs from arbitrary positions anyhow, and the current implementation doesn't update the chain head until everything is validated anyhow. So unless anyone can think of any other reason why this might be beneficial, I'm happy to drop this statement.